### ISSUE TITLE
CDN-hosted browser files

### ISSUE THREAD

**Issue by josephrocca** [kleisauke/wasm-vips#12]
Exciting project! Thanks for working on this and sharing it.

In [this comment](https://github.com/kleisauke/wasm-vips/issues/4#issuecomment-838408218) it is mentioned that the browser files can be downloaded from here:
* https://wasm-vips.kleisauke.nl/lib/vips.d.ts
* https://wasm-vips.kleisauke.nl/lib/vips.js
* https://wasm-vips.kleisauke.nl/lib/vips.wasm
* https://wasm-vips.kleisauke.nl/lib/vips.worker.js

Would it be possible to publish these in such a way that they can be accessed via unpkg.com or deno.land/x with a simple import like e.g. this?
```js
import vips from "https://unpkg.com/wasm-vips@0.0.2/web/vips.js"
```

**Comment by kleisauke** [comment]
As far as I know, unpkg has occasionally issues with the CORS headers, but jsDelivr should work without any issues. If you've set the [COOP/COEP headers](https://github.com/kleisauke/wasm-vips#browser) correctly, the upcoming v0.0.2 (RSN) release allows you do to this:
```html
<script crossorigin="anonymous" src="https://cdn.jsdelivr.net/npm/wasm-vips@0.0.2/lib/web/vips.js"></script>
<script type="module">
  const WASM_VIPS_CDN_BASE = 'https://cdn.jsdelivr.net/npm/wasm-vips@0.0.2/lib';
  let workerObjectUrl;

  const createWorkerURL = (name) => {
    if (workerObjectUrl) return workerObjectUrl;
    const blob = new Blob(
      [`importScripts('${WASM_VIPS_CDN_BASE}/web/${name}');`],
      {'type': 'application/javascript'}
    );
    workerObjectUrl = (window.URL ? URL : webkitURL).createObjectURL(blob);
    return workerObjectUrl;
  };

  const vips = await Vips({
    locateFile: name => name.endsWith(".worker.js") ? createWorkerURL(name) : `${WASM_VIPS_CDN_BASE}/${name}`
  });

  const im = vips.Image.black(100, 100);
  console.log(JSON.stringify({
    width: im.width,
    height: im.height,
    space: im.interpretation,
    channels: im.bands,
    depth: im.format
  }));
</script>
```
(If you want test it today, you can swap `https://cdn.jsdelivr.net/npm/wasm-vips@0.0.2` with `https://wasm-vips.kleisauke.nl`)

For Deno, I think you can do this in the upcoming release (i.e. just reuse Node's ES6 module):
```js
import Vips from "https://cdn.jsdelivr.net/npm/wasm-vips@0.0.2/lib/node-es6/vips.mjs";
const vips = await Vips();
```
(untested)

**Comment by josephrocca** [comment]
Ah cool! Any reason why you can't just change some relative imports within `lib/web/vips.js` so it's a one-line import statement / script tag for web browsers too? From your code snippet, it looks like a bit of a mess to import in the browser. I've seen libs that use workers and wasm (e.g. ONNX Web Runtime) work with a single script tag or import statement.

Also curious why the Deno 2-liner you showed wouldn't work in the browser?

**Comment by kleisauke** [comment]
The `locateFile` logic is only needed if you don't host wasm-vips yourself. It's needed to workaround a CORS issue with web workers, see:
https://stackoverflow.com/questions/25458104/can-should-html5-web-workers-use-cors-for-cross-origin

ONNX Web Runtime seems to handle this logic in their own wrapper.
https://github.com/microsoft/onnxruntime/blob/03181caeaef9343bd2e8d7999386893562ebb23a/js/web/lib/wasm/wasm-factory.ts#L120-L138

I don't think this is an option for wasm-vips, because that would break the `preRun` function, which is useful for things that need to happen before the program starts running (e.g. to set up environment variables, directories and/or preloaded files). The only simplification I could think of is to just merge the `lib/web/` directory with `lib/`, which I just did with commit fa502a2e2eb45667d29dcaef993a667d972a3c00.

I had a further look at Deno, but I think this won't work with wasm-vips Node's ES6 module, since that would require Node's `worker_threads` module.
https://github.com/emscripten-core/emscripten/blob/9a1d61ae7aef37e530f1e0ccb080b35844f37935/src/worker.js#L21
https://github.com/denoland/deno_std/tree/854756a85df1a219237bcce1fcb750025b2e6810/node#supported-modules

Although it's not an ES6 module, perhaps the web module works on Deno? I'll investigate that further before publishing v0.0.2.

**Comment by josephrocca** [comment]
@kleisauke Thanks for working to simplify this! If it works in the browser, it shouldn't be too hard to get it working in Deno - at this point a huge portion of the compatibility gap has been closed.

You could skip the script-tag-based imports and just do a Deno/browser ES6 module. If needed, users can always "emulate" the old (async) script tag import (i.e. import in non-module contexts) with a dynamic `import(...)` statement + an assignment to the `window` global:

```html
<script>import("https://cdn.jsdelivr.net/npm/wasm-vips@0.0.2/lib/vips.mjs").then(m => window.Vips=m.default);</script>
```
Another thought: Could that `locateFile` logic be included within the library itself so it's as simple as specifying an option in the constructor? Or even better, perhaps the script tag could just do some detection on where it has been downloaded from (with [document.currentScript](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript) and/or [import.meta](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import.meta)) and handle the `locateFile` stuff automatically?

**Comment by kleisauke** [comment]
I'm thinking about adding a `web-es6` directory, which involves a partial revert of commit https://github.com/kleisauke/wasm-vips/commit/ba6f7eec2d5adbadd1282fc41af43d901cece097. Hopefully that can be used on Deno as well. Producing ES6 modules on the web by default is not an option, as that would break compatibility with Firefox.

You're right about the `locateFile` logic could simply be included in the library itself. I just did this with commit https://github.com/kleisauke/wasm-vips/commit/0d776ceb41b089289b11047b6eb6b64e225e4ad3 and updated the previous usage comment and binaries hosted on https://wasm-vips.kleisauke.nl/.

Note that for the playground at https://wasm-vips.kleisauke.nl/playground/, you may need to visit https://wasm-vips.kleisauke.nl/clear-cache to make sure your browser's cache is cleared for this domain.

**Comment by kleisauke** [comment]
Commit https://github.com/kleisauke/wasm-vips/commit/02998f1d32d67de8ddd0052dba56241ef0221381 adds support for ES6 modules on the web. I decided to include this in a separate file (`vips-es6.js`), since a separate subdirectory for that is somewhat confusing.

Commit https://github.com/kleisauke/wasm-vips/commit/36f4f0e0cb72164e2b6c0add333d428d2d7ff560 ensures that this ES6 web module can also be used on Deno. I've updated the previous Deno usage (you can swap that URL with `https://wasm-vips.kleisauke.nl/lib/vips-es6.js`, if you want to test it today).

**Comment by josephrocca** [comment]
Brilliant - just tested on Deno and it works great!

**Comment by josephrocca** [comment]
Hey @kleisauke, is it possible to access v0.0.2 via a CDN url? E.g. something like:
```
import vips from "https://unpkg.com/wasm-vips@0.0.2/dist/vips.js"
```

**Comment by kleisauke** [comment]
Is this for use on the web? If so, ES6 module import doesn't currently work via CDNs due to:
https://github.com/kleisauke/wasm-vips/blob/372e5e345cf0c57b4866b44fad5568eac2823eef/src/CMakeLists.txt#L167

I can recommend using the CommonJS import syntax as described in https://github.com/kleisauke/wasm-vips/issues/12#issuecomment-1067001784. This would also ensure that your application would work on Firefox.

**Comment by josephrocca** [comment]
Oh, I forgot that you mentioned it'd be available via jsdelivr in your earlier comment, sorry! I assumed that I'd be able to browse the npm-published files here: https://unpkg.com/browse/wasm-vips@0.0.2 but for some reason unpkg gives a 404 for that page, so I assumed that there was some extra step involved to get the files from npm onto a CDN like unpkg/jsdelivr, but https://unpkg.com/wasm-vips@0.0.2/lib/vips.js loads fine, so I'm not sure what's going on there. I'll use the jsdelivr URL in any case. Sorry about that, and thanks!

**Comment by kleisauke** [comment]
FWIW, I had to make the CORS workaround (introduced in commit 0d776ceb41b089289b11047b6eb6b64e225e4ad3) opt-in with commit https://github.com/kleisauke/wasm-vips/commit/5a21668ffaa48e6ea0b3756b5c4b2934f57f568d.

> Edit: I wonder if it's worth replacing `<script src="vips.js"></script>` with `<script src="https://cdn.jsdelivr.net/npm/wasm-vips@0.0.2/lib/vips.js"></script>` in the readme?

Sorry, I just noticed this edit. I think it's better to leave it that way, since you'll have to set the `workaroundCors` setting now (I've updated comment https://github.com/kleisauke/wasm-vips/issues/12#issuecomment-1067001784 to reflect this).

**Comment by kleisauke** [comment]
> Is this for use on the web? If so, ES6 module import doesn't currently work via CDNs due to:
> https://github.com/kleisauke/wasm-vips/blob/372e5e345cf0c57b4866b44fad5568eac2823eef/src/CMakeLists.txt#L167

The `workaroundCors` setting is supported in ES6 modules with commit 16d462e015052d7fee397695837161a5ad5500d3 and will be available in v0.0.19.
